Sources/Datadog/URLSessionAutoInstrumentation/Interception/URLSessionInterceptor.swift

```diff
@@ -30,66 +30,87 @@ internal protocol URLSessionInterceptorType: class {
 }
 
 internal class URLSessionInterceptor: URLSessionInterceptorType {
-    private let firstPartyHostsFilter: URLFilter
-    private let queue: DispatchQueue
+    /// Filters first party `URLs` defined by the user.
+    private let firstPartyURLsFilter: FirstPartyURLsFilter
+    /// Filters internal `URLs` used by the SDK.
+    private let internalURLsFilter: InternalURLsFilter
 
-    private let tracingInterceptionHandler: URLSessionTracingHandlerType
+    /// Handles tracing `Span` creation for intercepted resources. `nil` if Tracing is disabled.
+    internal let tracingHandler: URLSessionTracingHandlerType?
+    /// Handles RUM Resource creation for intercepted resources. `nil` if RUM is disabled.
+    internal let rumResourceHandler: URLSessionRUMResourcesHandlerType?
 
     // MARK: - Initialization
 
-    convenience init(configuration: FeaturesConfiguration.URLSessionAutoInstrumentation) {
+    convenience init(
+        configuration: FeaturesConfiguration.URLSessionAutoInstrumentation,
+        dateProvider: DateProvider
+    ) {
         self.init(
-            firstPartyHostsFilter: URLFilter(
-                includedHosts: configuration.userDefinedFirstPartyHosts,
-                excludedURLs: configuration.sdkInternalHosts
-            ),
-            tracingInterceptionHandler: URLSessionTracingHandler()
+            configuration: configuration,
+            tracingHandler: configuration.instrumentTracing ? URLSessionTracingHandler() : nil,
+            rumResourceHandler: configuration.instrumentRUM ? URLSessionRUMResourcesHandler(dateProvider: dateProvider) : nil
         )
     }
 
     init(
-        firstPartyHostsFilter: URLFilter,
-        tracingInterceptionHandler: URLSessionTracingHandlerType
+        configuration: FeaturesConfiguration.URLSessionAutoInstrumentation,
+        tracingHandler: URLSessionTracingHandlerType?,
+        rumResourceHandler: URLSessionRUMResourcesHandlerType?
     ) {
-        self.firstPartyHostsFilter = firstPartyHostsFilter
-        self.tracingInterceptionHandler = tracingInterceptionHandler
+        self.firstPartyURLsFilter = FirstPartyURLsFilter(configuration: configuration)
+        self.internalURLsFilter = InternalURLsFilter(configuration: configuration)
+        self.tracingHandler = tracingHandler
+        self.rumResourceHandler = rumResourceHandler
         self.queue = DispatchQueue(label: "com.datadoghq.URLSessionInterceptor", target: .global(qos: .utility))
```

Comment:
it's out of scope but anyway, i'm not sure if `interceptor` needs its own queue: RUM has its own queue already and starting spans in different queues may have effects on `activeSpan`.
also i guess that makes its tests async.

is there a reason for using a private queue here?

Reply:
This `queue` synchronizes the access to:
```swift
/// Maps `URLSessionTask` to its `TaskInterception` object.
private var interceptionByTask: [URLSessionTask: TaskInterception] = [:]
```
While we can expect that single `URLSession` synchronizes its calls using its own `queue`, we must assume that the user can use several `URLSessions`.

Speaking on linking network `Spans` with the `activeSpan` running in the app - we're not really focusing on it:
- it was never an objective / requirement to do support it;
- this won't be soon possible (RUMM-696) as the `Span` will be generated by backend.

Reply:
if you prefer removing this queue/less queues less synchronization, we can use associated objects on `task`
```swift
func taskCreated(task) {
  task.interception = Interception(...)
}
func taskCompleted(task) {
  task.interception?.completed()
}
```
how does that look like?

Reply:
Indeed, that would mean no need for synchronisation. However, referring to the RFC, what we have discussed in CHALLENGE 4 was preparing the solution to not depend on runtime modifications. And with its current implementation the `URLSessionInterceptor` does not depend on any objc runtime at all. This means, we could easily make it `public` and let users call it directly, having all the functionality of RUM and APM instrumentation for `0` swizzling / associated objects cost.

We don't focus on making the `URLSessionInterceptor` public now, but given two comparable approaches, I'd rather prefer the one which scales better for CHALLENGE 4. WDYT?

Reply:
i think we can make it public in both cases because `task.interception` will be private, can be used within `Interceptor.swift` only ✅ 
i guess assoc objects are different than swizzling as we add new things and not replacing existing things, it's closer to `extension`s in that sense.
we don't change the memory layout of types or anything and also i don't think the cost of associated object is heavier than queue sync'ing.
sounds good?

Reply:
Although I see associated objects as a method of last resort, I did it a try. When going this way, the `testRandomlyCallingDifferentAPIsConcurrentlyDoesNotCrash()` test (in `URLSessionInterceptorTests`) shows the synchronization issue in `TaskInterception`:

<img width="884" alt="Screenshot 2020-10-08 at 08 34 03" src="https://user-images.githubusercontent.com/2358722/95423305-26002500-0941-11eb-8f8d-1e50e464a24b.png">

It's because `TaskInterception.register(metrics:)`, `TaskInterception.register(completion:)` and `TaskInterception.isDone` are called from two separate interceptor methods:
```swift
// URLSessionInterceptor

func taskMetricsCollected(urlSession: URLSession, task: URLSessionTask, metrics: URLSessionTaskMetrics)
func taskCompleted(urlSession: URLSession, task: URLSessionTask, error: Error?)
```
and this test tries to run them concurrently for the same task.

The [`URLSession.init()` documentation says](https://developer.apple.com/documentation/foundation/urlsession/1411597-init) however that each session synchronizes its delegate cals and completion handlers using session `queue`, so this test should be rather improved to synchronize calls for the same `URLSession` instance or removed entirely.

However, it is fact that calls between different session instances will happen concurrently, meaning that we may need to add thread synchronization as soon as we introduce any shared resource in `URLSessionInterceptor` or any of it's child components: `URLSessionTracingHandler` or `URLSessionRUMResourcesHandler`. If we're OK with this eventual cost, I can change to use runtime utils instead of synchronization `queue`. LTMWDYT @buranmert @xgouchet @mariusc83

Reply:
to my understanding tasks are intercepted independently, that's why i can't think of a case with shared resource among interceptions. can you think of any?
if there are, then yes we can't get rid of `queue`, otherwise associated objects can simplify the code.

if `TracingHandler` or `RUMHandler` change `Tracer`'s or `RUMMonitor`'s states, those handlers can have their own queue but still `Interceptor` can stay without queue.

Reply:
conclusion: we keep the shared resource and async execution instead of using `associatedObject`s
@ncreated 's point is AO is in `objc/runtime.h` and considered dangerous, it's also slow.
i don't agree and i think it would improve the code in our case but it works either way, so convo is resolved ✅